caper/resource_analysis.py

```diff
@@ -0,0 +1,338 @@
+import fnmatch
+import json
+import logging
+from abc import ABC, abstractmethod
+from collections import defaultdict
+
+from .cromwell_metadata import CromwellMetadata, convert_type_np_to_py
+from .dict_tool import flatten_dict
+
+logger = logging.getLogger(__name__)
+
+
+class ResourceAnalysis(ABC):
+    """
+    Class constants:
+        DEFAULT_REDUCE_IN_FILE_VARS:
+            Function to be used for reducing x vector.
+            e.g. sum, min, max, ...
+        DEFAULT_TARGET_RESOURCES:
+            Keys to make y vector.
+    """
+
+    DEFAULT_REDUCE_IN_FILE_VARS = sum
+    DEFAULT_TARGET_RESOURCES = ('stats.max.mem', 'stats.max.disk')
+
+    def __init__(self, metadata_jsons):
+        """Solves y = f(x) in a statistical way where
+        x is a vector of input file sizes (e.g. size_each([bam, bowtie2_index_tar, ...]))
+        y is a vector of resources (e.g. [max_mem, max_disk, ...])
+
+        self._tasks is extended (across all workflows) list of resource monitoring
+        result from gcp_monitor():
+        [
+            {
+                'task_name': TASK1,
+                'workflow_id': WORKFLOW_ID_OF_TASK1,
+                ...
+            },
+            {
+                'task_name': TASK2,
+                'workflow_id': WORKFLOW_ID_OF_TASK2,
+                ...
+            }
+        ]
+        Args:
+            metadata_jsons:
+                List of metadata JSON file URIs or metadata JSONs
+                or CromwellMetadata objects.
+        """
+        self._tasks = []
+        for metadata_json in metadata_jsons:
+            if isinstance(metadata_json, CromwellMetadata):
+                cm = metadata_json
+            else:
+                cm = CromwellMetadata(metadata_json)
+            self._tasks.extend(cm.gcp_monitor())
```

Comment:
This has side effects yeah? I think it would be appropriate here to make this into an `@property`. The `__init__` could just store the `metadata_jsons`, then you can have an `@property` called `tasks` that instantiates the `CromwellMetadata` and populates `_tasks`.

Reply:
Renamed `self._tasks` to `self._task_resources`.

`cm.gcp_monitor()` is very expensive to be called in a `@property`.
It recursively parses each meta JSON and all input files in it.

I agree that this shouldn't be called in `__init__()`.
So I added `collect_resource_data()` which explicitly parses meta JSONs to collect resource data.